### PR TITLE
Fanotify support

### PR DESCRIPTION
**Pull request by amir73il**
* Build alternative tools fsfnotifywait and fsnotifywatch
* The alternative tools auto detect and use fanotify instead of inotify on new kernels (>=v5.9)
* The alternative tools enable watching an entire filesystem without setting up recursive watches
* Setting up fanotify filesystem watches require admin privileges

### CODE REVIEW

File: .clang-format
Comment (ericcurtin) [inline .clang-format:4]: Could `BreakBeforeBraces: Linux` be removed please, the others can stay if you'd like.

**Comment by ericcurtin** [comment]
In general, this stuff looks great by the way!

File: .clang-format
Comment (amir73il) [reply .clang-format:4]: OK. I also added:
+IndentCaseLabels: false
+PointerAlignment: Right
+AllowShortIfStatementsOnASingleLine: WithoutElse
+PenaltyBreakBeforeFirstCallParameter: 100
+PenaltyBreakAssignment: 100
+PenaltyExcessCharacter: 10

And fixed all my changes to conform to this style.
However, I excluded src/inotifywait.c and src/inotifywatch.c from style check 
because they are space indented and I didn't want to change that.

**Review by ericcurtin — COMMENTED** [review]
Happy enough to merge this for support on Linux, if we are not confident on FreeBSD support, we can ifdef it out for needs be.

Things that need to be removed before merge:
- gcc specific stuff like: _attribute__((unused))
- src/test_demo.sh
- edits to clang-format file (I don't want to get into the business of maintaining a custom .clang-format file, not worth it)

Things required before merging:
- a couple of automated tests to ensure this code gets some runtime during out builds
- -h print_help stuff
- man pages

File: .clang-format
Comment (amir73il) [reply .clang-format]: @ericcurtin are you fine with me leaving just tab indentation as your first comment? Because in a later comment you requested to remove all edits from this file.

File: libinotifytools/src/inotifytools.c
Comment (amir73il) [reply libinotifytools/src/inotifytools.c]: Removed it by making the var non static

**Comment by amir73il** [comment]
> Happy enough to merge this for support on Linux, if we are not confident on FreeBSD support, we can ifdef it out for needs be.
> 
> Things that need to be removed before merge:
> 
> * gcc specific stuff like: _attribute__((unused))
done

> * src/test_demo.sh
done

> * edits to clang-format file (I don't want to get into the business of maintaining a custom .clang-format file, not worth it)
Left only Tab Indent per your earlier comment (and because it fits most of the code)
 
> 
> Things required before merging:
> 
> * -h print_help stuff
> * man pages

I suppose you mean --writes and --timeout changes I made for the demo?
I removed those from the MR.
--fanotify and --global already have -h and man pages update, let me
know if you meant more than that.

> * a couple of automated tests to ensure this code gets some runtime during out builds

Can you help me get started with sharness?
What is the expected result if the kernel is not new enough (v5.9) to support the
fanotify modes needed by the code?
Is there an example for a test that runs depending on kernel version or some
available feature?

**Comment by ericcurtin** [comment]
> Can you help me get started with sharness?

build_and_test.sh script should give you an idea how to run them. Example scripts in t directory

> What is the expected result if the kernel is not new enough (v5.9) to support the
fanotify modes needed by the code?
> Is there an example for a test that runs depending on kernel version or some
available feature?

You could check kernel version using `uname -r`, I would probably do something like:

```
if [ -e /sys/kernel/slab/fanotify_event ]; then
  # Do test stuff
fi
```

The build servers may not be a recent as 5.9. But, I notice on ubuntu 20.04, it seems to run 5.4 with fanotify support, it may have been backported, that should be good enough for our needs.

**Comment by amir73il** [comment]
> > Can you help me get started with sharness?
> 
> build_and_test.sh script should give you an idea how to run them. Example scripts in t directory
> 
> > What is the expected result if the kernel is not new enough (v5.9) to support the
> > fanotify modes needed by the code?
> > Is there an example for a test that runs depending on kernel version or some
> > available feature?
> 
> You could check kernel version using `uname -r`, I would probably do something like:
> 
> ```
> if [ -e /sys/kernel/slab/fanotify_event ]; then
>   # Do test stuff
> fi
> ```
>

That's is not enough, but that is not what I meant.
Sure, I can test kernel version explicitly, but the question is what is the
expectation from the test in that case.
Should the result be the same as if test has passed or is there any special way to
indicate to the harness that test prerequisites are not met?
I am familiar with LTP and xfstests infrastructures, which do have this option
that is why I am asking.

> The build servers may not be a recent as 5.9. But, I notice on ubuntu 20.04, it seems to run 5.4 with fanotify support, it may have been backported, that should be good enough for our needs.


Alas, v5.4 is not good enough.
The code on branch fanotify_name_fid relies on FAN_REPORT_DFID_NAME mode,
which was added in v5.9.
Even though the code on branch fanotify_fid should work with kernel v5.4, the resulting
report does not include information that users of inotify-tools are use to get from inotify,
so I am reluctant to request a merge for that code alone.

**Comment by ericcurtin** [comment]
Had a few meetings.

```
if grep -q "#define.*FAN_REPORT_FID" /usr/include/linux/fanotify.h; then
  # do test
fi
```

Checks like above are ok if they work. Thanks for pointing out the 5.9 requirement for this new code. I'm gonna try a put together a build server for that that in Travis or somewhere else in parallel to help you out.

**Comment by ericcurtin** [comment]
> I'm gonna try a put together a build server for that that in Travis or somewhere else in parallel to help you out.

Clarifying this is gonna be a local fedora machine with recent kernel.

**Comment by ericcurtin** [comment]
> Sure, I can test kernel version explicitly, but the question is what is the expectation from the test in that case.

Within the test, run the relevant test cases if the currently running kernel can support it, otherwise skip.

**Comment by amir73il** [comment]
> > Sure, I can test kernel version explicitly, but the question is what is the expectation from the test in that case.
> 
> Within the test, run the relevant test cases if the currently running kernel can support it, otherwise skip.

OK. But there is another issue. Even with recent kernel --fanotify requires admin privleges.
Can the test run as root user?
I can add [ $(id -u) == 0 ] as prerequisite for the test, but if none of the build servers
execute the tests as root user, the tests will never run in CI...
I could run the test using sudo if I know that the build server user can execute
sudo without a password prompt...

**Comment by amir73il** [comment]
Pushed a new version with improved documentation in man page about kernel version and privileges requirements for fanotify.
Also improved the human readable warning when running inotifywait --fanotify with missing kernel support or no privelges.
Wrote a simple test as a reference design - it executes inotifywait --fanotify and examines stderr to determine if test should be run or skipped. On kernel v5.9, when tests is run as root user it passes. Otherwise, it does not run.

**Comment by ericcurtin** [comment]
We are gonna have to remove the clang-format on/offs , we will need more tests in other files to test more than just was the exit code 0. Please move:

```
  /*
   * Assuming sys/fanotify.h does exist, but linux/fanotify.h doesn't have
   * FAN_REPORT_FID.
   */
  #include <sys/fanotify.h>
```

to the top of the file in libinotifytools/src/inotifytools/fanotify-nosys.h

The rest looks reasonably ok. I have a fedora machine ready with a 5.11 kernel to test. For the root user check, whatever works, I might change it after sometime, but we need some more tests here.

**Comment by amir73il** [comment]
> We are gonna have to remove the clang-format on/offs

OK.

>  we will need more tests in other files to test more than just was the exit code 0.

OK.

> Please move:
> 
> ```
>   /*
>    * Assuming sys/fanotify.h does exist, but linux/fanotify.h doesn't have
>    * FAN_REPORT_FID.
>    */
>   #include <sys/fanotify.h>
> ```
> 
> to the top of the file in libinotifytools/src/inotifytools/fanotify-nosys.h
>

OK.
 
> The rest looks reasonably ok. I have a fedora machine ready with a 5.11 kernel to test. For the root user check, whatever works, I might change it after sometime

This I did not understand.
I am asking what my options are because I do not know the testing infrastructure.

* Can I run sudo in the test script or will it prompt for password and hang the test?
* Alternatively, can you make sure that on the 5.11 kernel all the tests run as root user? (this is how I written the sample test - it checks it if was run by root and if it wasn't the test is skipped)

>  but we need some more tests here.

Sure. I just need to get passed the root user issue first, so I can test the sample test on the fedora machine.

**Comment by ericcurtin** [comment]
assume the test is run like `sudo ./test.t` Within the test check that the feature exists or kernel version (whatever you prefer) is high enough so we don't break our current builds

**Comment by amir73il** [comment]
> assume the test is run like `sudo ./test.t` Within the test check that the feature exists or kernel version (whatever you prefer) is high enough so we don't break our current builds

Oh, that's done already in the sample test. To clarify, I implemented the helper:
`fanotify_supported() {
    ../../src/inotifywait --fanotify 2>&1 | grep -q 'No files specified'
}`

The way it works is that inotifytools_init() is called before failing on "No files specified to watch!" and inotifytools_init(fanotify) verifies that fanotify is both supported by kernel and that user is root and warns with hints in warn_inotify_init_error() if conditions are not met.

So you meant that you have setup a fedora machine for your own tests? Now I get it. I though you were talking about the GitHub CI. Sure, I will prepare more tests that won't run on old kernels or not root and pass on new kernel and root.

**Comment by ericcurtin** [comment]
> So you meant that you have setup a fedora machine for your own tests? Now I get it. I though you were talking about the GitHub CI. Sure, I will prepare more tests that won't run on old kernels or not root and pass on new kernel and root.

Yes it's a little early for this kernel in CI, but in time you will see this code tested in CI. Until that point, I'll test every PR on my vm to ensure no breakages.

**Comment by ericcurtin** [comment]
I'm actually starting to think long-term it might be better to have fanotifywait, fanotifywatch binaries to prevent overloaded command-line args, things like:

`Watch entire filesystem of any directories passed as arguments. Implies...`

seem okish short-term, but it's important to remember once something is generally committed to trunk here we tend to suffer from Hyrum's Law and I get emails whenever something changes. So short-term doesn't really apply in our main branch.

Would you be happy with inotifywatch/inotifywait --fanotify --global long-term, like in 10 years? Because once it's committed this way I doubt we will ever change it and create fanotifywait/fanotifywatch binaries. Creating fanotifywait/fanotifywatch binaries also helps make it blatantly obvious that we now support fanotify which I'd like people to be aware of. Might be enough of a kick for the distro-maintainers to update their inotifytools rpms/debs/etc. What do you think?

**Comment by amir73il** [comment]
> Would you be happy with inotifywatch/inotifywait --fanotify --global long-term, like in 10 years? Because once it's committed this way I doubt we will ever change it and create fanotifywait/fanotifywatch binaries. Creating fanotifywait/fanotifywatch binaries also helps make it blatantly obvious that we now support fanotify which I'd like people to be aware of. Might be enough of a kick for the distro-maintainers to update their inotifytools rpms/debs/etc. What do you think?

I like the idea of fanotifywait/fanotifywatch binaries.
* We won't build them for __FreeBSD__
* I don't think we need to clone inotifywatch.c/inotifywait.c
* We could let same binary behave as --fanotify when installed in different name, but I think rebuild with define LINUX_FANOTIFY would be better for differentiating the command line args and help 
* I suppose we'd need to clone the man pages

**Comment by amir73il** [comment]
CC @brauner @jankara 

On second thought, if we are going to re-brand intofify-tools, I don't think we should use the name "fanotify" as it means very little to end users and they should not really care about the backend driver name.

Very few people would understand the difference between inotifywait and fanotifywait and indeed there is not much difference (yet). The major difference from users POV is the option to watch entire filesystem or entire mount.

For that reason, I think that we better re-brand to fsnotifywatch/fsnotifywait.
fsnotify* tools will auto select the backend driver according to kernel support, user privileges and requested feature (e.g. --filesystem, --mount, --recursive) and there will also be a "low level" option to use --inotify --fanotify to force selection of the backend driver for testing or whatnot.

Furthermore, fsnotifywatch --subtree can auto use --filesystem or --mount and filter events by subtree in userspace, which is a facility not yet available in the kernel. fsnotifywatch --subtree can also fall back to plain old --recursive, with its known limitations.

**Comment by ericcurtin** [comment]
> On second thought, if we are going to re-brand intofify-tools, I don't think we should use the name "fanotify" as it means very little to end users and they should not really care about the backend driver name.

I would rather not re-brand inotify-tools, as I alluded to above I try to maintain compatibility for everyone. The pkg, binary, etc. have always been prefixed inotify. If we want to add binaries with different names we can, it doesn't break anybodies usecase.

I'm actually ok with the command line args in this pull request or the fanotify* binaries that we don't build on FreeBSD, etc. But whatever in merged in this PR will not be changed. Pointing out the obvious, but changing command-line args and/or binary names to fsnotifywatch/fsnotifywait are the types of changes that lead to angry email.

**Comment by ericcurtin** [comment]
> fanotify_supported() { ../../src/inotifywait --fanotify 2>&1 | grep -q 'No files specified' }

Great way of checking in the end by the way.

**Comment by amir73il** [comment]
> > On second thought, if we are going to re-brand intofify-tools, I don't think we should use the name "fanotify" as it means very little to end users and they should not really care about the backend driver name.
> 
> I would rather not re-brand inotify-tools, as I alluded to above I try to maintain compatibility for everyone. The pkg, binary, etc. have always been prefixed inotify. If we want to add binaries with different names we can, it doesn't break anybodies usecase.
> 
> I'm actually ok with the command line args in this pull request or the fanotify* binaries that we don't build on FreeBSD, etc. But whatever in merged in this PR will not be changed. Pointing out the obvious, but changing command-line args and/or binary names to fsnotifywatch/fsnotifywait are the types of changes that lead to angry email.

You miss-understood me.
I did not mean that we should change inotify* to fsnotify*.
I meant that instead of adding new binaries fanotify* we will add fsnotify* binaries
and that instead of the fanotify* binaries that imply fanotify backend, we will install new fsnotify* binaries that auto detect the optimal backend to use (but can also be forced to use an explicit backend)

**Comment by brauner** [comment]
> CC @brauner @jankara
> 
> On second thought, if we are going to re-brand intofify-tools, I don't think we should use the name "fanotify" as it means very little to end users and they should not really care about the backend driver name.
> 
> Very few people would understand the difference between inotifywait and fanotifywait and indeed there is not much difference (yet). The major difference from users POV is the option to watch entire filesystem or entire mount.
> 
> For that reason, I think that we better re-brand to fsnotifywatch/fsnotifywait.

Sounds good. As long as the manpage explains the possible drivers and users have the option to select their driver of choice explicitly that's perfect.

**Comment by ericcurtin** [comment]
> and that instead of the fanotify* binaries that imply fanotify backend, we will install new fsnotify* binaries that auto detect the optimal backend to use (but can also be forced to use an explicit backend)

Apologies for miss-understanding. If we want to create inotify/fsnotify/fanotify binaries, seems fine to me. The project will remain called inotify-tools (no re-brands please). There are details I'm unsure of like does fsnotify end up doing an exec(inotify) or exec(fsnotify) or do we use a shared library. 6 main function/man pages/etc. to maintain. I can live with this I'm sure we can do it in a nice way.

On a personal note, hope all is well @brauner long-time no chat :)

**Comment by amir73il** [comment]
> Apologies for miss-understanding. If we want to create inotify/fsnotify/fanotify binaries, seems fine to me. 

No fanotify* only legacy inotify* and new fsnotify* binaries

> The project will remain called inotify-tools (no re-brands please). 

Agreed.

> There are details I'm unsure of like does fsnotify end up doing an exec(inotify) or exec(fsnotify) or do we use a shared library. 

In any case libinotifytools in common.
Two options for the binaries:
1. fsnotifywatch is the same binary as intotifywatch installed in different name and detects its mode by checking argv[0] 
2. fsnotifywatch is built by compiling intotifywatch.c with -D LINUX_FANOTIFY

Will try and see which is easier to implement/maintain.

> 6 main function/man pages/etc. to maintain.

so just 4.

**Comment by ericcurtin** [comment]
> * fsnotifywatch is the same binary as intotifywatch installed in different name and detects its mode by checking argv[0]
> * fsnotifywatch is built by compiling intotifywatch.c with -D LINUX_FANOTIFY

Separate binaries, option 1 is tempting, but I fear we may break someones workflow, someone who renames binaries, does some symlinking etc. We could fallback to inotify* versions of course. In this case we would have to support --fanotify forever I feel like some people may want to rename binaries (well my bigger fear is that they are doing this already).

So, option 2 please, and if we could just use the existing define SYS_FANOTIFY_H_EXISTS_AND_WORKS (you can rename that if you want to LINUX_FANOTIFY, because it's not in our main branch), that would be great.

**Comment by ericcurtin** [comment]
When we do get this completed I think it's time for a 4.x release

**Comment by amir73il** [comment]
> Please move:
> 
> ```
>   /*
>    * Assuming sys/fanotify.h does exist, but linux/fanotify.h doesn't have
>    * FAN_REPORT_FID.
>    */
>   #include <sys/fanotify.h>
> ```
> 
> to the top of the file in libinotifytools/src/inotifytools/fanotify-nosys.h
> 

But the idea is that the installed linux/fanotifyf.h is too old.
This file replaces the old linux/fanotify.h and does not replace sys/fanotify.h.
That is why this is at the bottom.

Do you want to rename the file suffix -nosys and macro name SYS_FANOTIFY_H_EXISTS_AND_WORKS
to something like LINUX_FANOTIFY_H_EXISTS_AND_WORKS to make that clear?

**Comment by ericcurtin** [comment]
Is it possible to do this in this file instead libinotifytools/src/inotifytools/fanotify.h.in Something like this (or just handle it in this file in general)? Want to keep include blocks together

```
#ifndef _INOTIFYTOOLS_FANOTIFY_H
#define _INOTIFYTOOLS_FANOTIFY_H 1

/* this is defined to 1 if <sys/fanotify.h> works. */
#undef SYS_FANOTIFY_H_EXISTS_AND_WORKS

#ifndef SYS_FANOTIFY_H_EXISTS_AND_WORKS
#include "fanotify-nosys.h"
#endif // SYS_FANOTIFY_H_EXISTS_AND_WORKS

#include <sys/fanotify.h>

#endif
```

**Comment by amir73il** [comment]
> > * fsnotifywatch is the same binary as intotifywatch installed in different name and detects its mode by checking argv[0]
> > * fsnotifywatch is built by compiling intotifywatch.c with -D LINUX_FANOTIFY
> 
> Separate binaries, option 1 is tempting, but I fear we may break someones workflow, someone who renames binaries, does some symlinking etc. We could fallback to inotify* versions of course. In this case we would have to support --fanotify forever I feel like some people may want to rename binaries (well my bigger fear is that they are doing this already).
> 
> So, option 2 please, and if we could just use the existing define SYS_FANOTIFY_H_EXISTS_AND_WORKS, that would be great.

I pushed a version that builds fsnotifywatch/fsnotifywait. Currently, they will only be built on system with SYS_FANOTIFY_H_EXISTS_AND_WORKS, although we could make that also an opt-in configure parameter --enable-fsnotify if we want to allow building fsnotify* on an older system.

The man pages are shared with the inotify* man pages so we do not need to maintain duplicate pages.
Please say if you like how this looks and then I think all that will remain in TODO are the tests. Right?

**Comment by ericcurtin** [comment]
Sounds good to me, a couple of tests should be all we need to get this into the main from this point. We may end up duplicating the man pages in the end, but this can do for now.

**Comment by ericcurtin** [comment]
A FUSE filesystem of some sort might be useful for testing

**Comment by amir73il** [comment]
Why FUSE? For unprivileged user mount?
FUSE happens to be one of the few fs not supported by fanotify report fid mode.
We can test with loop device ext4 for now

**Comment by ericcurtin** [comment]
> We can test with loop device ext4 for now

Sounds good to me.

**Comment by amir73il** [comment]
> Sounds good to me, a couple of tests should be all we need to get this into the main from this point. We may end up duplicating the man pages in the end, but this can do for now.

With current version on kernel >=v5.9, (cd t; sudo make) will run and pass all the tests with and without --fanotify.
The test found two --fanotify bugs (with --no-dereference and %w format print) and I fixed them.

FYI, with upcoming v5.13-rc1, sudo won't be needed to run fanotify tests.

There are no tests for --global yet

**Comment by ericcurtin** [comment]
Need to get rid of new options (--fanotify, --global) in inotify* binaries, that's the point of the new ones I guess.

**Comment by amir73il** [comment]
> Need to get rid of new options (--fanotify, --global) in inotify* binaries, that's the point of the new ones I guess.

Done.

Updated man pages and tests to reflect that change.

Also added the configure option --enable-fanotify for building fsnotify* on older system.

I took the liberty to rename --global => --filesystem because 'global' is a bit ambiguous
and it will be even more ambiguous if we even want to add support for fanotify mount watches.

**Comment by ericcurtin** [comment]
fsnotify*/fanotify* (call what you like) binaries should support fanotify only

inotify binaries should support inotify only

other than that I think we are there.

**Comment by ericcurtin** [comment]
I ran build_and_test.sh (putting sudo before the integration test part of that script):

```
*** inotifywait-no-dereference-respond-to-symlink-event.t ***
*** inotifywait-no-event-occured.t ***
not ok 1 - Exit code 0 is returned
not ok 1 - Exit code 2 is returned
#	
#	    run_ inotifywait
#	
#	
#	    test_expect_code 2 run_ inotifywait
#	
# failed 1 among 1 test(s)
1..1
make[1]: *** [Makefile:38: inotifywait-no-dereference-respond-to-symlink-event.t] Error 1
make[1]: *** Waiting for unfinished jobs....
# failed 1 among 1 test(s)
1..1
make[1]: *** [Makefile:38: inotifywait-no-dereference-ignore-symlinked-file.t] Error 1
```

Tests fail. Known issue?

**Comment by amir73il** [comment]
> fsnotify*/fanotify* (call what you like) binaries should support fanotify only

Why?

fsnotify* is about the functionality not about the driver.
It is a super set of inotify*.
Users should not care about the driver only about the functionality.
When users requests functionality fsnotify* can auto select the optimal driver to perform this functionality with the current kernel support and user capabalities.


> 
> inotify binaries should support inotify only
> 
> other than that I think we are there.

**Comment by amir73il** [comment]
> I ran build_and_test.sh (putting sudo before the integration test part of that script):
> 
> ```
> *** inotifywait-no-dereference-respond-to-symlink-event.t ***
> *** inotifywait-no-event-occured.t ***
> not ok 1 - Exit code 0 is returned
> not ok 1 - Exit code 2 is returned
> #	
> #	    run_ inotifywait
> #	
> #	
> #	    test_expect_code 2 run_ inotifywait
> #	
> # failed 1 among 1 test(s)
> 1..1
> make[1]: *** [Makefile:38: inotifywait-no-dereference-respond-to-symlink-event.t] Error 1
> make[1]: *** Waiting for unfinished jobs....
> # failed 1 among 1 test(s)
> 1..1
> make[1]: *** [Makefile:38: inotifywait-no-dereference-ignore-symlinked-file.t] Error 1
> ```
> 
> Tests fail. Known issue?

No. Tests passed for me.
Please try to run 
`cd t; sudo TEST_OPTS=-v make`

**Review by ericcurtin — COMMENTED** [review]
inotify-tools was/is always about the driver, it's what distinguishes the project from other tools. We don't do things like traditional Unix polling of a directory. It's one of the reasons why people choose us over alternative generic file watch/wait tools. inotify* binaries will not be relegated to a legacy binary that supports a subset of the functionality, because the inotify* binaries are what people use at present. No re-brands. If you want to create another binary for fanotify* purposes that makes sense to have a clean interface. I don't see any auto switching to optimal backend here, I just see a build time switch. It's important to remember that people statically link (this broke recently it was noticed very quickly #129 #131), so the binaries could be deployed anywhere.

That said if you want to create a parent super binary after this PR separately, if you really need it and it solves a use-case for you. I will reluctantly merge it in a separate PR and maintain it. inotify* binaries will be supported for the foreseeable future as a 1st class non-legacy binary.

The tests look great and they pass for me, thanks very much for that! I had a permissions issue on my vm I had to solve.

**Comment by amir73il** [comment]
> inotify-tools was/is always about the driver, it's what distinguishes the project from other tools. We don't do things like traditional Unix polling of a directory. It's one of the reasons why people choose us over alternative generic file watch/wait tools. inotify* binaries will not be relegated to a legacy binary that supports a subset of the functionality, because the inotify* binaries are what people use at present. No re-brands. If you want to create another binary for fanotify* purposes that makes sense to have a clean interface. I don't see any auto switching to optimal backend here, I just see a build time switch. It's important to remember that people statically link (this broke recently it was noticed very quickly #129 #131), so the binaries could be deployed anywhere.
> 
> That said if you want to create a parent super binary after this PR separately, if you really need it and it solves a use-case for you. I will reluctantly merge it in a separate PR and maintain it. inotify* binaries will be supported for the foreseeable future as a 1st class non-legacy binary.
> 

I am confused about this comment because the way I understand your reservations do not fit my the current code. Is this comment referring to the commit "Opt-in to use fanotify in place of inotify"? Because the following commit "Build fsnotifywait and fsnotifywatch binaries" reverts all change of behavior to inotify* tools as you requested and allows the use of explicit backend choice --fanotify/--inotify ONLY to the fsnotify* tools, and the man page is also updated to reflect that in commit "Create man page entries for fsnotifywait.1 and fsnotifywatch.1". 

If it is important to you I can reorder the commits to avoid this mid-series change of behavior to inotify* tools, but I was hoping that won't be necessary.

For now, I just removed the "fallback to inotify" behavior which is controversial.
fsnotify* tools always use fanotify unless --inotify was specified explicitly.
Needless to say, intotify* tools always use inotify. Period.

**Review by ericcurtin — APPROVED** [review]
This looks good for now, it is obvious which interface they are using now, without no defaults changing depending on kernel version built on, which is nice. Thanks for completing this.

**Comment by amir73il** [comment]
> This looks good for now, it is obvious which interface they are using now, without no defaults changing depending on kernel version built on, which is nice. Thanks for completing this.

Thank you for your feedback and patience :)

**Comment by ericcurtin** [comment]
Something noticed in the build, haven't looked into it yet:

./sharness.sh: line 763: 10663 Terminated              ../../src/$* --monitor --quiet --outfile $logfile --format '%w%f%0' --no-newline --event create --event moved_to --event moved_from $(realpath ./)  (wd: ~)

**Comment by ericcurtin** [comment]
There's still work here, but it can be worked on from this point.

@amir73il @jankara @brauner  will fanotify require admin or other special privileges for the foreseeable future?

fsnotify* binaries requiring sudo by default isn't the most ideal. That will prevent some existing users ever using fanotify with these binaries (I might add a hint for that in the output when fanotify fails, defer to inotify).

I might reach out to existing users just to see what they would prefer to see as the default here.

**Comment by amir73il** [comment]
> @amir73il @jankara @brauner will fanotify require admin or other special privileges for the foreseeable future?

Kernel commit 7cea2a3c505e ("fanotify: support limited functionality for unprivileged users") is already in kernel v5.13-rc1.
If you run the tests without sudo on latest rc kernel all tests are expected to run and pass except for the single --filesystem test that won't run.

--filesystem will forever require admin privileges.
Right now it requires admin privileges in the init user ns.
Going forward that will be relaxed for watching filesystems that were mounted inside userns, like FUSE, tmpfs, overlayfs.
Users will need admin privileges in the relevant user ns in those cases.

**Comment by ericcurtin** [comment]
Thanks @amir73il, I was peaking at that commit earlier, just making sure, thanks for your efforts.

**Pull request merged**
